### PR TITLE
452 stat block parser fixes

### LINKED ISSUE #452: Stat block parser blows the stack when parsing extremely complicated stat blocks
## Bug Description

Trying to parse a stat block from "The New Age" results in errors.

## Steps to Reproduce

1. Copy stat block from the PDF:
```
Hope
Hapless protomatter Steward, age 28
(use Hope’s stats for any Steward)
STR 26 CON 23 DEX 15 INT 11 POW 10 CHA 14
HP 19 WP 10 SAN 0
SKILLS: Art (Photography) 53%, Athletics 93%,
Computer Science 24%, Craft (Electronics) 64%,
Craft (Lockpick) 74%, Disguise 74%, Dodge 38%,
Drive 52%, Persuade 47%, SIGINT 61%, Stealth 92%,
Unarmed Combat 44%, Unnatural 14%.
ARMOR: Stewards are unharmed by all melee and
unarmed blows. Firearms have all damage reduced
to 1, and Hope regenerates 1D4 HP per round. Fire,
electricity, acid, chemical irritants, and hypergeometric
rituals inflict damage normally and stop all
regeneration for the remainder of combat.
ATTACKS: Slam and Strangle 44%, damage 2D6.
Grapple 44%, see “Draw Close.”
Swallow, damage special, see “Swallow.”
Kiss, damage special, see “Kiss.”
Mi-go electric gun 27%, Lethality 2%, 15%, or 25%,
ignores armor.
ELECTRIC GUN: This weapon appears as a pulsing,
softball-sized glob with three dangling tendrils. When
the weapon is picked up, the tendrils wrap around
the user’s forearm. The weapon is a biomechanical organism that can eject bolts of electrical current, sort
of like an electric eel. The charge of lightning emitted
from the weapon is variable, but Hope has no need
for anything but the deadliest setting. The only reason
she would turn down the charge would be to protect
her beloved Valiant, but in such instances, she gives
her master the gun and attempts to suffocate or rend his
enemies up close. The electric gun ignores body armor
but can be blocked by cover.
DRAW CLOSE: After a successful grapple, a victim must
succeed on a STR×5 check opposing Hope’s successful
attack roll to break free. On a failure, they suffer either
Swallow or Kiss the next round.
KISS: Hope spurts protomatter down the grappled victim’s
throat and takes control of their body. Any Agent
subjected to the Kiss loses control of their character
until Hope is destroyed or withdraws her parasitic
protomatter. The victim loses 1/1D6 SAN due to
helplessness.
SWALLOW: After successfully grappling a victim, Hope
liquifies herself, surrounds, and engulfs the body in
a constricting protomatter sac. Hope can then inflict
2D6 damage to the victim with each action without an
attack roll. The victim must take an action and roll STR×5
to break free.
SAN LOSS: 1/1D6 if seen as protomatter.
```
2. Parse and receive the following error:
```
extract-attacks.js:60 Uncaught (in promise) RangeError: Maximum call stack size exceeded
    at Array.join (<anonymous>)
    at extractAttacksImpl (extract-attacks.js:60:62)
    at extractAttacksImpl (extract-attacks.js:77:12)
    at extractAttacksImpl (extract-attacks.js:77:12)
    at extractAttacksImpl (extract-attacks.js:77:12)
    at extractAttacksImpl (extract-attacks.js:77:12)
    at extractAttacksImpl (extract-attacks.js:77:12)
    at extractAttacksImpl (extract-attacks.js:77:12)
    at extractAttacksImpl (extract-attacks.js:77:12)
    at extractAttacksImpl (extract-attacks.js:77:12)

```

## Expected Behavior

Expected the block to be parsed into an Actor.

## System Info

- Foundry VTT Version: 14.364
- Delta Green System Version: 2.0.1
- Access Method: [x] Browser / [ ] Electron App
- OS & Browser (if applicable): 
```
Google Chrome | 151.0.7922.108 (Official Build) (arm64)
-- | --
Revision | 4744b886309d987d292e43232776d2206cccb13d-refs/branch-heads/7922@{#2606}
OS | macOS Version 26.5.2 (Build 25F84)
JavaScript | V8 15.1.206.16
```

## Additional Context

Image

### PR DESCRIPTION
**Pull request by jiblits**
## Checklist

- [ ] This is meant for the next release.
- [x] This is meant for a hotfix.
- [ ] This is a Build System change.
- [ ] This needs more reviewers than normal
- [ ] This intentionally introduces regressions that will be addressed later.
- [ ] The change will require updates to the documentation.
- [ ] Please do not send commits here without coordinating closely with the owner.

## What does this PR do?

Really complicated statblocks from operations such as The New Age cause the parser to fail to parse or to get into infinite
loops. This addresses those issues.

**This PR is not complete. There is another example stat block that needs to be tested before this is ready to merge**

## How to Test

1. Tests are automated, but the actor statblock parser can be used to test different stat blocks

## Related Issue(s)

Closes #452

### CODE REVIEW

File: module/macros/stat-parser-macro.js [hunk 1]
```diff
@@ -358,21 +369,23 @@ async function RegexParseNpcStatBlock(inputStr, actorType) {
...
+    const [attackSkill, killRadius, expense] = ["custom", "N/A", "N/A"];
     const [range, skillMod] = [0, 0];
     const skillTarget = skillModifier;
     const equipped = true;
     const isLethal = damage === undefined && lethality !== undefined;
     await newActor.AddWeaponItemToSheet(
       name,
       description,
-      damage,
+      damage ?? "",
```
Comment (jiblits) [inline module/macros/stat-parser-macro.js:388]: Fun note: When `damage` was undefined, which could be the case when dealing with attacks that are lethality only, the attack would default to 1d8.

File: test/macros/stat-block-parser.test.js [hunk 2]
```diff
@@ -384,8 +425,274 @@ describe("ParseStatBlock", () => {
...
+            name: "Unarmed",
+            skillModifier: 73,
+            damage: "1d4",
+            notes: "(1d8)",
+          },
+          {
+            name: "Too complicated to parse. See notes.",
+            notes:
+              "(after the meeting [see page 38] , valiant uses no weapons but his hands and his psychic powers",
+          },
```
Comment (jiblits) [inline test/macros/stat-block-parser.test.js:632]: I am not certain how to handle the following input

```
(After THE MEETING [see page 38], Valiant uses no
weapons but his hands and his psychic powers.)
```

so I basically just have the parser exclaim that the entry couldn't be parsed and dump the entire attack into the notes section

File: test/macros/stat-block-parser.test.js [hunk 3]
```diff
@@ -384,8 +425,274 @@ describe("ParseStatBlock", () => {
...
+          {
+            name: "Knife (marine Ka-bar)",
+            skillModifier: 57,
+            damage: "1d6+1",
+            notes: "(1d6+3) armor piercing 3",
+          },
+          {
+            name: "S&w Model",
```
Comment (jiblits) [inline test/macros/stat-block-parser.test.js:640]: This also runs into some issues where we are accidentally taking the following input and confusing what the skill value is:

```
S&W Model 645 Pistol 62%, damage 1D10.
```

With the simpler stat blocks, there weren't gun-nut style stat blocks, just simpler names like "Assault Rifle" which is why this is getting caught incorrectly. The original implementation went about removing the `%` symbols because it made it easier, but perhaps we bring them back so we can much more easily tell when something is the skill modifier. There's enough tests here that we should be able to make changes and know that the outputs remain consistent.

File: test/macros/statblocks/alien-steward-stats.txt
Comment (jiblits) [inline test/macros/statblocks/alien-steward-stats.txt:1]: These tests are using official DG statblocks. I don't know if there's a better way around this since it's easier to copy the examples of the input we are going to be getting from users to replicate the issue.

Now that I have tests passing I can do super basic rewording to strip out the copyright, but it's a lot of the little things that caused a lot of the problems (using the open-close style quotes, enemy special attacks like Kiss, etc.)

File: module/macros/stat-parser-macro.js [hunk 4]
```diff
@@ -350,6 +350,17 @@ async function RegexParseNpcStatBlock(inputStr, actorType) {
     actorData.system.notes = GetNotesFromInput(inputStr);
   }
 
+  if (actorType === "unnatural") {
+    const {
+      sanloss: { failedLoss, successLoss, notes },
+    } = statBlock;
```
Comment (camrun91) [inline module/macros/stat-parser-macro.js]: Unnatural import crashes when the stat block has no SAN LOSS section. Megalomaniac has no `loss:` line, so `statBlock.sanloss` is undefined. This block destructures it with no guard.

Add a guard. Only copy sanity fields when `statBlock.sanloss` is set, or default before destructuring.

Repro: import the Megalomaniac stat block as an unnatural NPC.

File: module/macros/stat-block-parser.js [hunk 5]
```diff
@@ -143,7 +180,7 @@ function extractAttributesImpl(tokens, accumulator) {
 
 function extractSkillsImpl(tokens, skillsAccum, skillNameAccum) {
   if (tokens === []) {
```
Comment (camrun91) [inline module/macros/stat-block-parser.js]: Empty skills blow the stack. `if (tokens === [])` compares object references. An empty array never matches, so `extractSkillsImpl` recurses with no base case.

Use `if (tokens.length === 0)`.

Repro: parse a stat block with an empty skills section.

File: module/macros/stat-block-parser.js [hunk 6]
```diff
@@ -104,14 +137,16 @@ export function determineNextState(token) {
...
+      return States.SanLoss;
     default:
       return States.Unknown;
   }
 }
 
 function extractAttributesImpl(tokens, accumulator) {
-  if (tokens === []) {
-    return [accumulator, tokens];
+  if (tokens.length === []) {
```
Comment (camrun91) [inline module/macros/stat-block-parser.js]: This empty check never runs. `tokens.length === []` compares a number to a new array, so it is always false.

Use `if (tokens.length === 0)`.

File: module/macros/extract-attacks.js [hunk 7]
```diff
@@ -85,16 +89,26 @@ function extractAttacksImpl(tokens, accumulatedAttacks, incompleteAttack) {
...
       .map(capitalize)
       .join(" ");
-    partialAttack.skillModifier = skillModifier;
+    partialAttack.skillModifier = Number.isNaN(skillModifier)
+      ? 0
+      : skillModifier;
     return extractAttacksImpl(rest, accumulatedAttacks, partialAttack);
   }
 
+  if (accumulatedAttacks.length === 0) {
```
Comment (camrun91) [inline module/macros/extract-attacks.js]: This early return stops parsing with tokens still on the stack. Knife AP stays in `notes` instead of `armorPiercing`. `S&W Model 645` gets skillModifier 645 from the model number. The Conway fixture in `test/macros/stat-block-parser.test.js` (lines 633-643) expects these wrong values.

Remove the early return and keep parsing trailing tokens. Parse `armor piercing N` after damage. Do not treat a bare integer after the weapon name as the skill modifier when more attack tokens follow.

Repro: parse the Conway stat block attacks section.

File: module/macros/stat-block-parser.js [hunk 8]
```diff
@@ -360,27 +411,39 @@ function parseStatBlockImpl(tokens, state, statBlock) {
...
+    return parseStatBlockImpl(otherLines, nextState, {
       ...statBlock,
       ...disordersAndAdaptations,
     });
   }
 
-  return parseStatBlockImpl(rest, nextState, statBlock);
+  if (state === States.SanLoss) {
+    const [sanLossTokens] = groupEntriesUntilNextSection(tokens.slice(1));
+    const sanloss = ExtractSanLoss(sanLossTokens);
```
Comment (camrun91) [inline module/macros/stat-block-parser.js]: Parse throws when SAN LOSS has no entry. `groupEntriesUntilNextSection` can return an empty group. Then `ExtractSanLoss` calls `.split("/")` on undefined.

Skip extraction when `sanLossTokens` is missing or empty, or guard inside `ExtractSanLoss` before split.

Repro: parse an unnatural block with a `SAN LOSS:` header and no value on that line.

File: module/macros/stat-parser-macro.js (hunk 4 above)
Comment (camrun91) [inline module/macros/stat-parser-macro.js:356]: Unnatural import crashes when the stat block has no SAN LOSS section. Megalomaniac has no `loss:` line, so `statBlock.sanloss` is undefined. This code destructures `statBlock.sanloss` with no guard.

Add a guard. Copy sanity fields only when `statBlock.sanloss` is set. Or set a default before you destructure.

Repro: import the Megalomaniac stat block as an unnatural NPC.

File: module/macros/stat-block-parser.js (hunk 5 above)
Comment (camrun91) [inline module/macros/stat-block-parser.js:182]: Stack overflow on empty skills. `if (tokens === [])` compares object references. An empty array never matches, so `extractSkillsImpl` recurses with no base case.

Use `if (tokens.length === 0)`.

Repro: parse a stat block with an empty skills section.

File: module/macros/stat-block-parser.js (hunk 6 above)
Comment (camrun91) [inline module/macros/stat-block-parser.js:148]: This empty check never runs. `tokens.length === []` compares a number to a new array, so it is always false.

Use `if (tokens.length === 0)`.

File: module/macros/extract-attacks.js (hunk 7 above)
Comment (camrun91) [inline module/macros/extract-attacks.js:98]: This early return stops parsing with tokens still on the stack. Knife AP stays in `notes` instead of `armorPiercing`. `S&W Model 645` gets skillModifier 645 from the model number. The Conway fixture in `test/macros/stat-block-parser.test.js` (lines 633-643) expects these wrong values.

Remove the early return and keep parsing trailing tokens. Parse `armor piercing N` after damage. Do not treat a bare integer after the weapon name as the skill modifier when more attack tokens follow.

Repro: parse the Conway stat block attacks section.

File: module/macros/stat-block-parser.js (hunk 8 above)
Comment (camrun91) [inline module/macros/stat-block-parser.js:435]: Parse throws when SAN LOSS has no entry. `groupEntriesUntilNextSection` can return an empty group. Then `ExtractSanLoss` calls `.split("/")` on undefined.

Skip extraction when `sanLossTokens` is missing or empty. Or guard inside `ExtractSanLoss` before split.

Repro: parse an unnatural block with a `SAN LOSS:` header and no value on that line.

**Review by camrun91 — APPROVED** [review]
My human review did not find anything. I found some technical things that should be validated using Cursor. do with that what you want.

File: module/macros/extract-attacks.js (hunk 7 above)
Comment (jiblits) [reply module/macros/extract-attacks.js:98]: This looks like it might be fabricating something. I don't have a "Conway stat block" in the tests so don't see how I could reproduce it.

Also dropping it results in dropping special abilities. I don't know how perfect we should expect the parser to get because we're dealing with _very_ unstructured text that gets really weird with the really special NPCs vs. the chaff like Security Officers and whathaveyou.

File: module/macros/stat-block-parser.js (hunk 6 above)
Comment (jiblits) [reply module/macros/stat-block-parser.js:148]: This was a legit bug / oversight. Will be fixing this one.

File: module/macros/stat-block-parser.js (hunk 5 above)
Comment (jiblits) [reply module/macros/stat-block-parser.js:182]: Again another oversight that will be addressed in another PR.

File: module/macros/stat-block-parser.js (hunk 8 above)
Comment (jiblits) [reply module/macros/stat-block-parser.js:435]: I guess because we are dealing with any input we should probably address this, but are there any scenarios where we'd be dealing with the unnatural that doesn't have San Loss?

It can be addressed because users can be goofs sometimes, but it's an edge case that shouldn't really happen, right?

File: module/macros/stat-parser-macro.js (hunk 4 above)
Comment (jiblits) [reply module/macros/stat-parser-macro.js:356]: Megalomaniac isn't an unnatural NPC, so in a normal scenario this wouldn't happen. I think we just address the scenario where it's missing and leave it at that.

**Review by jiblits — COMMENTED** [review]
Here's my take of the Cursor feedback:
- Conway comment - fabrication/error by the LLM
  - Note: There's known issues where parsing very complicated entries (like extremely specific guns) that is difficult to make sense of with the current parsing approach.
- Some were oversights that can easily be caught and addressed with tests (the array checks)
- The Unnatural one is valid but a bit of an edge case IMO. Maybe just make sure that if there's no unnatural info we return an object with a valid shape so callers further down don't crash.
